### PR TITLE
Modify LZW score adjustment heuristic for long reads to reduce the number of long, low-complexity sequences

### PR DESCRIPTION
**Pull request by katrinakalantar**
# Description

Joe's LZW threshold was optimized for a single read length (150 bp). We added a linear extrapolation based on read length but it is letting too many reads through, causing rapsearch to choke. We should improve the extrapolation function. Requires careful benchmarking across a range of read lengths. 

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix a feature that knowingly causes existing functionality to not work as expected)

# Test and Reproduce

1. LZW scores were calculated for IDseq samples with a range of read lengths (see plot for avg. read length x avg. lzw score many samples. A model was fit to these points. The resulting model formula was used to generate an adjustment parameter.

```
Call:
lm(formula = y_orig ~ log(x_orig))

Residuals:
      Min        1Q    Median        3Q       Max
-0.043178 -0.009059 -0.000343  0.005155  0.082182

Coefficients:
             Estimate Std. Error t value Pr(>|t|)
(Intercept)  1.043456   0.020121   51.86   <2e-16 ***
log(x_orig) -0.113148   0.004001  -28.28   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.01952 on 61 degrees of freedom
Multiple R-squared:  0.9291,	Adjusted R-squared:  0.928
F-statistic: 799.7 on 1 and 61 DF,  p-value: < 2.2e-16
```

lzw_model

2. The fix was tested on several samples, including two samples with long reads (> 300nt). The results were comparable, with slightly more reads removed in the LZW stage post-fix.

**Reads filtered by LZW**
Sample1: (pre-fix) 108 reads, (post-fix) 652 reads
Sample2: (pre-fix) 196 reads, (post-fix) 1114 reads

# Checklist:

- [ ] I have run through the testing script to make sure current functionality is unchanged
- [x] I have done relevant tests that prove my fix is effective or that my feature works
- [x] I have spent time testing out edge cases for my feature
- [ ] I have updated the test script or pull request template if necessary
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: idseq_dag/steps/run_lzw.py [hunk 1]
```diff
@@ -84,14 +94,17 @@ def lzw_score(sequence, threshold_readlength):
 
         if seq_length > threshold_readlength:
             # Make sure longer reads don't get excessively penalized
-            adjustment_heuristic = (1 + (seq_length - threshold_readlength) / 1000) # TODO: revisit
-            score = lzw_fraction * adjustment_heuristic
+            #adjustment_heuristic = (1 + (seq_length - threshold_readlength) / 1000) # TODO: revisit
```
Comment (tfrcarvalho) [inline idseq_dag/steps/run_lzw.py]: Any particular reason to keep these commented?

File: idseq_dag/steps/run_lzw.py [hunk 2]
```diff
@@ -51,8 +52,17 @@ def count_reads(self):
         self.should_count_reads = True
         self.counts_dict[self.name] = count.reads_in_group(self.output_files_local()[0:2])
 
+    # predict LZW score from sequence length based on model fit to mean LZW score across mean lengths
+    # model formula: lm(formula = y ~ log(x))
```
Comment (tfrcarvalho) [inline idseq_dag/steps/run_lzw.py]: Sorry, probably newbie question... what data (`x` and `y`) was used to get to these values? Did you just use all the data in our db? Want to make sure we can get back to them from the information here.

**Comment by katrinakalantar** [comment]
Time comparison on staging:
**Sample1** - 1 hour, 54 minutes (LZW branch) v. 2 hours, 4 minutes (master)
**Sample2** - 2 hours, 24 minutes (LZW branch) v. 2 hours, 15 minutes (master)

File: idseq_dag/steps/run_lzw.py (hunk 2 above)
Comment (tfrcarvalho) [reply idseq_dag/steps/run_lzw.py]: I just read the explanation in the description better. Would still be good to clear state how to retrieve those values from the DB e.g. `select all lengths from DB for a particular set of hosts...`

File: idseq_dag/steps/run_lzw.py (hunk 1 above)
Comment (katrinakalantar) [reply idseq_dag/steps/run_lzw.py]: No, good point - I will uncomment and remove the previous function.

File: idseq_dag/steps/run_lzw.py (hunk 2 above)
Comment (katrinakalantar) [reply idseq_dag/steps/run_lzw.py]: The x-values were `mean read length` for sequences from a sample, and the y-values were `mean lzw score` for the same sample. This was done on a subset of samples along a defined distribution.

Specifically, eight sequence length windows were defined and within each window, a maximum of ten samples were selected. For the lower and higher windows, data may not have been available to reach the ten samples. The distribution of selected samples were as follows: 

- 25 - 60bp (2 samples)
- 60 - 80bp (10 samples)
- 80 - 130 bp (10 samples)
- 130 - 180bp (10 samples)
- 180 - 220bp (10 samples)
- 220 - 280bp (10 samples)
- 280 - 500bp (10 samples)
- > 500 bp (1 sample)

Across a total of 63 samples, the average read length and LZW compression score were computed for the first 500 reads. From these values, a linear model was fit using the R stats package.

Comment (tfrcarvalho) [reply idseq_dag/steps/run_lzw.py]: Is there some of this info that we should include in the comments in the code?
This conversation might be lost after you merge.

Comment (katrinakalantar) [reply idseq_dag/steps/run_lzw.py]: Added the following comment text:
Across a total of 63 samples with varying read lengths, the average read length and LZW compression score were computed for the first 500 reads. From these values, a linear model was fit using the R stats package.
